### ISSUE TITLE
Production Ready?

### ISSUE THREAD

**Issue by cia48621793** [javascript-obfuscator/javascript-obfuscator#35]
I'm about to JSObfuscator in my SPA project. 

But I'm not sure if **this** is **'secure'** enough to keep malicious client away from **stealing/exploiting/attacking my code**.

By now, only main business logic is obfuscated. Vendor packages, for performance's sake, are not. High obfuscation profile is chosen, and still I'm very afraid of **decompiliation/deobfuscation**...

_Is JSObfuscator battle-ready to defend the code beneath?_

**Comment by sanex3339** [comment]
Hi, you can't restore original variable/function names, but you can restore original strings, even if you use `high obfuscation profile`. Question is - "how much time you can spend for deobfuscation"? Because deobfuscation of `high obfuscation profile` takes significant amount of time.

In `0.9.0` (will released this weekend) new feature will added - `controlFlowFlattening`, take a look here
https://github.com/javascript-obfuscator/javascript-obfuscator#controlflowflattening

So, i can recommend to wait until this weekend and use `0.9.0` (or you can use latest beta version now `npm i javascript-obfuscator@0.9.0-beta.5`

**Comment by sanex3339** [comment]
Just take a look at obfuscated angular2 (medium obfuscation, so few strings not obfuscated)
https://gist.github.com/sanex3339/cf5fe8289051cf9ef1a54d6e000b7364

**Comment by cia48621793** [comment]
＠sanex3339 That is a lot of strings...My browser is almost frozen.

I attempted to decompile the obfuscated code at [jsnice](http://jsnice.org/)...

**Comment by sanex3339** [comment]
You cannot decompile this strings with jsnice, becouse they are encoded with `rc4` encoding.

**Comment by cia48621793** [comment]
@sanex3339 No. Around 10% of the original code is retrieved.

**Comment by sanex3339** [comment]
You mean strings? In the example with angular 2 about 25% of strings was not affected by obfuscation, because `stringArrayThreshold` is 0.75.

**Comment by cia48621793** [comment]
I think JSObfuscator can be a minimal protection against hackers, given that my code will be delivered to client and end there, I get no control of it, hackers can still, at the end of the day, dump and crack and attack my code, but it guess that would take long.

However, hackers can still read the DOM tree, which means, the great hard work of my artists/illustrators/designers can be stolen in no time. But that is hard to prevent.

PSA: [react-armor](https://github.com/elierotenberg/react-armor) can be a good combination with JSObfuscator.

**Comment by sanex3339** [comment]
You should know what you never can protect your JS code with guarantee what it will be saved.
JavaScript - is interpreted programming language, and only way to defend it - make maximum transformation of it without loosing runtime stability.

Some methods can look like very good obfuscation, for example http://www.iteral.com/jscrush/
But i can get original code in 30 seconds, it's just visible protection.

`javascript-obfuscator` (do not confuse with https://github.com/caiguanhao/js-obfuscator) makes various amount of transformation over original source code like: rename identifiers, collecting all strings into array (in raw or encoded format), this array can be rotated, control flow flattening.

If someone will try to debug this transformations, debug protection can help.

But this is still JS code, and after N time you can find `debug protection code`, remove it and after that you can start to debug obfuscated code. For getting all original values of encoded strings you can spend N time (depends on code size), but after this you will get original values.

So, in conclusion, you never can obfuscate your JS code and say "It's under 100% protection". It's just impossible with JS.
